docs/proposals/templates.md

```diff
@@ -0,0 +1,400 @@
+    // Items using ${PARAMETER_NAME}
+    Name string
+
+    // Optional: The name that will show in UI instead of parameter 'Name'
+    DisplayName string
+
+    // Optional: Parameter can have description
+    Description string
+
+    // Optional: Value holds the Parameter data. If specified, the generator
+    // will be ignored. The value replaces all occurrences of the Parameter
+    // ${Name} expression during the Template to Config transformation.
+    Value string
+
+    // Optional: Generate specifies the generator to be used to generate
+    // random string from an input value specified by From field. The result
+    // string is stored into Value field. If empty, no generator is being
+    // used, leaving the result Value untouched.
+    Generate string
+
+    // Optional: From is an input value for the generator.
+    From string
+
+    // Optional: Indicates the parameter must have a value.  Defaults to false.
+    Required bool
+}
+```
+
+As seen above, parameters allow for metadata which can be fed into client implementations to display information about the 
+parameter’s purpose and whether a value is required.
+
+The value of the parameter can be explicitly defined (this should be considered a default value for the parameter, clients 
+which process templates are free to override this value based on user input), or determined via a defined `Generator`.  
+Generators are named by a string, and can be optionally supplied an input parameter via the `From` field.
+
+For example:
+
+```
+{
+      "name": "GITHUB_WEBHOOK_SECRET",
+      "description": "A secret string used to configure the GitHub webhook",
+      "generate": "expression",
+      "from": "[a-zA-Z0-9]{40}"
+}
+```
+
+This parameter will utilize the expression generator, which takes as an input a regex style string and will generate a 
+random value which conforms to the provided regex pattern.
+
+Generators are defined by an interface, so [new generators can easily be added](https://github.com/openshift/origin/tree/master/pkg/template/generator):
+```
+type Generator interface {
+    GenerateValue(expression string) (interface{}, error)
+}
+```
+
+### API Endpoints
+
+* **/processedTemplates** - when a template is POSTed to this endpoint, all parameters in the template are processed, values 
+generated as necessary, and substituted into appropriate locations in the object definitions.  Validation is performed to ensure
+required parameters have a value (either supplied in the original template or generated by a defined generator).  In addition 
+labels defined in the template are applied to the object definitions.  Finally the customized template (still a `Template` object) 
+is returned to the caller.
+
+### Workflow
+
+#### Template Instantiation
+
+Given a well-formed template, a client will
+
+1. Optionally set an explicit `value` for any parameter values the user wishes to explicitly set (eg non-generated parameters for 
```

Comment:
this could handled via a separate endpoint which took a new object consisting of a parameter map and a template name/template reference.

it would retrieve the template, set the parameter values, and submit to the existing endpoint for substitution.

(or the existing endpoint could be extended to take this new object type and do the above work).

thoughts?